### ISSUE TITLE
define-model-declaration

### ISSUE THREAD

**Issue by dword-design** [vuejs/eslint-plugin-vue#3069]
**Please describe what the rule should do:**

Analogous to [define-props-declaration](https://eslint.vuejs.org/rules/define-props-declaration), there should be a rule define-model-declaration that enforces a type-based declaration.

**What category should the rule belong to?**

[x] Enforces code style (layout)
[ ] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**

```vue
<script setup lang="ts">
defineModel();
</script>
```

```vue
<script setup lang="ts">
defineModel({ type: String });
</script>
```

**Workaround**
```ts
'no-restricted-syntax': [
  'error',
  {
    message:
      'defineModel() must use an explicit type parameter, e.g. defineModel<string>().',
    selector:
      "CallExpression[callee.name='defineModel']:not([typeArguments]):not([typeParameters])",
  },
],
```

**Comment by FloEdelmann** [comment]
Or maybe `define-props-declaration` should also check `defineModel`? Most/all users who prefer TypeScript-based prop definitions will also prefer TypeScript-based model definitions. But including it in a rule called `define-props-declaration` may be counterintuitive.

**Comment by dword-design** [comment]
Looks like [require-prop-types](https://eslint.vuejs.org/rules/require-prop-types) also covers the model. So having the model as part of define-props-declaration would make sense to me.

I mean, in the end the question is if the model is considered props. Afaik stuff like defineModel is convenience above props, so on the lower level the model is also props. So yeah, happy to reframe it to "add model to define-props-declaration". I can adjust the issue if we want to go this direction.
